azure/services/securitygroups/securitygroups.go

```diff
@@ -83,7 +95,55 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	// We go through the list of security groups to reconcile each one, independently of the result of the previous one.
 	// If multiple errors occur, we return the most pressing one.
 	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
-	for _, nsgSpec := range specs {
+	for _, resourceSpec := range specs {
+		nsgSpec, ok := resourceSpec.(*NSGSpec)
+		if !ok {
+			return errors.New("cannot convert network security group spec")
+		}
+
+		newAnnotation := map[string]interface{}{}
+
+		// We only want to update annotations for the security rules that are always managed by CAPZ.
+		for _, securityRuleSpec := range nsgSpec.SecurityRulesSpecs {
+			if _, alwaysManaged := securityRuleLastAppliedAnnotation[securityRuleSpec.Annotation]; !alwaysManaged {
+				continue
+			}
+			newAnnotation[securityRuleSpec.SecurityRule.Name] = securityRuleSpec.SecurityRule
+		}
+
+		// Retrieve the last applied security rules for all NSGs.
+		lastAppliedSecurityRulesAll, err := s.Scope.AnnotationJSON(azure.SecurityRuleLastAppliedAnnotation)
+		if err != nil {
+			return err
+		}
+
+		// Retrieve the last applied security rules for this NSG.
+		lastAppliedSecurityRules, ok := lastAppliedSecurityRulesAll[nsgSpec.Name].(map[string]interface{})
+		if !ok {
+			lastAppliedSecurityRules = map[string]interface{}{}
+		}
+
+		// Delete the security rules were removed from the spec.
+		for ruleName := range lastAppliedSecurityRules {
+			if _, ok := newAnnotation[ruleName]; !ok {
+				_, err := s.client.DeleteRule(ctx, nsgSpec, ruleName)
```

Comment:
Ah I wasn't aware PUT actually deletes unspecified rules. Is this separate from CreateOrUpdateResource? The existing implementation of that does not delete rules that aren't specified.

Reply:
The regular NSG PUT would be calling the [`CreateOrUpdateAsync` function](https://github.com/willie-yao/cluster-api-provider-azure/blob/update-securitygroups/azure/services/securitygroups/client.go#L83) correct? The existing code already calls this on each reconcile with the updated rules list, but the rules are not deleted in that case. I think they would have to be explicitly deleted if that functions is what you're referring to.

Reply:
^ Disclaimer that I'm not 100% sure it works like that so you might need to experiment a bit to see if this is even possible.

Reply:
Yeah, that's what CreateOrUpdateAsync does. I think the reason items aren't being deleted now is because `Parameters` checks "do all of the specified rules exist?" and only adds ones that are missing that should be there. All the rules that already exist on the resource in Azure get passed through as-is, but I wonder if deleting some from the list that we return in `Parameters` would also delete them in Azure. I expect they would based on how I understand HTTP PUT to generally work.

Reply:
Oh wow I never noticed that. With this discovery, I can pretty much throw out this whole PR in favor of detecting a change within `Parameters()`, as annotations are no longer needed. @CecileRobertMichon would this make sense? Since the parameters are only going to have user-specified security rules, all of those rules would be managed by CAPZ anyways.

Reply:
I believe the main reason we can't do this is that CAPZ doesn't have full ownership of the NSG, the azure cloud-provider also adds its own rules to expose services and users may have company Azure Policies that enforce some rules to be there, so CAPZ only "owns" the rules it previously created

Reply:
That makes sense! I did check which rules are being tracked by `Parameters()`, and they are only showing the rules from the user spec. Since it has both the existing rules and the newly applied rules from a spec, I don't think it should be interfering with cloud-provider's rules, unless they are being applied via the spec.

Reply:
I think I'm not following... How would you know which ones from existing rules need to be removed and which ones should be kept?

Reply:
I think I'm definitely misunderstanding something too.. My idea was that if the there is a rule in the [existing rules](https://github.com/willie-yao/cluster-api-provider-azure/blob/update-securitygroups/azure/services/securitygroups/spec.go#L70) that is no longer in the [current rules](https://github.com/willie-yao/cluster-api-provider-azure/blob/update-securitygroups/azure/services/securitygroups/spec.go#L71), we can set `update` to true and that rule would be deleted as a part of `CreateAndUpdate`

Reply:
How would you know which rules in existing rules (which is _ALL_ the current rules in the existing Azure Security Group resource) were previously created by CAPZ and therefore needs to be deleted (not included in the PUT) and which ones were created by something else (and should be preserved and included in the PUT)?

Reply:
@nojnhuh is this what you were thinking or are we missing something?

Reply:
@nojnhuh Since we need to distinguish what rules are previously created by CAPZ and need to be deleted, I think that we still have a need for the annotation in the type? Otherwise, I'm not sure how we would determine which rules are owned by CAPZ or not. https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3591#discussion_r1204390794